src/Qir/Execution/Tools/Executable/QirExecutableGenerator.cs

```diff
@@ -0,0 +1,92 @@
+        private readonly ILogger logger;
+
+        public QirExecutableGenerator(IClangClient clangClient, ILogger logger)
+        {
+            this.clangClient = clangClient;
+            this.logger = logger;
+        }
+
+        public async Task GenerateExecutableAsync(FileInfo executableFile, DirectoryInfo sourceDirectory, IList<DirectoryInfo> libraryDirectories, IList<DirectoryInfo> includeDirectories, IList<string> linkLibraries)
+        {
+            // Wrap in a Task.Run because FileInfo methods are not asynchronous.
+            await Task.Run(async () =>
+            {
+                var binDirectory = executableFile.Directory;
+                logger?.LogInfo($"Creating binary directory at {binDirectory.FullName}.");
+                executableFile.Directory.Create();
+
+                // Copy all library contents to bin.
+                logger?.LogInfo("Copying library directories into the executable's folder.");
+
+                if (!sourceDirectory.Exists)
+                {
+                    logger?.LogWarning($"Cannot find source directory: {sourceDirectory.FullName}");
+                }
+
+                var libDirs = new List<string>();
+                foreach (var dir in libraryDirectories)
+                {
+                    if (!dir.Exists)
+                    {
+                        logger?.LogWarning($"Cannot find given directory: {dir.FullName}");
+                    }
+                    else
+                    {
+                        CopyDirectoryContents(dir, binDirectory);
+                        libDirs.Add(dir.FullName);
+                    }
+                }
+
+                var includeDirs = new List<string>();
+                foreach (var dir in includeDirectories)
+                {
+                    if (!dir.Exists)
+                    {
+                        logger?.LogWarning($"Could not find given directory: {dir.FullName}");
+                    }
+                    else
+                    {
+                        includeDirs.Add(dir.FullName);
+                    }
+                }
+
+                var inputFiles = sourceDirectory.GetFiles().Select(fileInfo => fileInfo.FullName).ToArray();
+                await clangClient.CreateExecutableAsync(inputFiles, linkLibraries.ToArray(), libDirs.ToArray(), includeDirs.ToArray(), executableFile.FullName);
+            });
+        }
+
+        private void CopyFileIfNotExists(FileInfo fileToCopy, DirectoryInfo destinationDirectory)
+        {
+            var newPath = Path.Combine(destinationDirectory.FullName, fileToCopy.Name);
+            if (!File.Exists(newPath))
+            {
+                var newFile = fileToCopy.CopyTo(newPath);
+                logger?.LogInfo($"Copied file {fileToCopy.FullName} to {newFile.FullName}");
+            }
+        }
+
+        private void CopyDirectoryContents(DirectoryInfo directoryToCopy, DirectoryInfo destinationDirectory)
```

Comment:
This uses CopyFileIfNotExists so copies files only if destination file doesn't exist. Suggest reflecting this in the name. Just by the name I'd assume it will overwrite in this situation.

Reply:
Changed the name to reflect that it only copies the contents if it does not already exist in the target directory,